doc/reference/trd104-syscalls.md

```diff
@@ -794,7 +814,52 @@ If an exit syscall is successful, it does not return. Therefore, the return
 value of an exit syscall is always `Failure`. `exit-restart` and
 `exit-terminate` MUST always succeed and so never return.
 
-5 libtock-c Userspace Library Methods
+
+5.0 Upcall Types
+=================================
+
+This section defines the Upcalls that are valid for the kernel to invoke
+on userspace. Applications are responsible for ensuring that references
+to Upcalls passed to the kernel adhere to the ABI.
+
+5.1 The Null Upcall
+---------------------------------
+
+The Tock kernel defines an upcall pointer as the Null Upcall.
+The Null Upcall denotes an upcall that the kernel will never invoke.
+The Null Upcall is used for two reasons. First, a userspace process
+passing the Null Upcall as the upcall pointer for Subscribe
+indicates that there should be no more upcalls. Second, the first
+time a userspace process calls Subscribe for a particular upcall,
+the kernel needs to return upcall and application pointers indicating
+the current configuration; in this case, the kernel returns the Null
+Upcall. The Tock kernel MUST NOT invoke the Null Upcall.
+
+The Null Upcall upcall pointer MUST be 0x0. This means it is not possible
+for userspace to pass address 0x0 as a valid code entry point. Unlike
+systems with virtual memory, where 0x0 can be reserved a special meaning, in
+microcontrollers with only physical memory 0x0 is a valid memory location.
+It is possible that a Tock kernel is configured so its applications
+start at address 0x0. However, even if they do begin at 0x0, the
+Tock Binary Format for application images mean that the first address
+will not be executable code and so 0x0 will not be a valid function.
+In the case that 0x0 is valid application code and where the
+linker places an upcall function, the first instruction of the function
+should be a no-op and the address of the second instruction passed
+instead.
+
+If a userspace process invokes subscribe on a driver ID that is not
+installed in the kernel, the kernel MUST return a failure with an
+error code of `NODEVICE` and an upcall of the Null Upcall.
+
+
+5.2 SubscribableUpcall
```

Comment:
Yes I think we need this. I was looking for this and I'm not sure we define anywhere in this document how an upcall is encoded.